digdag-core/src/main/java/io/digdag/core/agent/ExtractArchiveWorkspaceManager.java

```diff
@@ -22,19 +31,53 @@
     public ExtractArchiveWorkspaceManager(TempFileManager tempFiles)
     {
         this.tempFiles = tempFiles;
+
     }
 
     @Override
     public <T> T withExtractedArchive(TaskRequest request, ArchiveProvider archiveProvider, WithWorkspaceAction<T> func)
             throws IOException
     {
-        try (TempDir workspacePath = createNewWorkspace(request)) {
-            Optional<StorageObject> in = archiveProvider.open();
-            if (in.isPresent()) {
-                ProjectArchives.extractTarArchive(workspacePath.get(), in.get().getContentInputStream());
+        TempDir workspacePath = null;
+        try {
+            try {
+                // A temporary workspace should be created for each try
+                // And always close the temporary workspace
+                workspacePath = RetryExecutor.retryExecutor()
```

Comment:
Let's add a comment here like....

> Here has retrying because there were some observations that downloading of S3 caused too early EOF. S3Storage uses ResumableInputStream but it can't rescue from too early end of streams. Because ProjectResource.putProject validates the file, getting IOException here is always unexpected.